src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -415,19 +417,28 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                 }
             }
 
+            object? instance;
             try
             {
-                return Activator.CreateInstance(type);
+                instance = Activator.CreateInstance(type);
             }
             catch (Exception ex)
             {
                 throw new InvalidOperationException(SR.Format(SR.Error_FailedToActivate, type), ex);
             }
+
+            if (instance is null
+                && Nullable.GetUnderlyingType(type) is { } nullableUnderlyingType)
```

Comment:
This is technically new functionality. Previously, null mutable structs that were initially null would end up nooping in BindNonScalar. In my mind it makes more sense to treat these the same whether they start as null or not.

Reply:
This makes sense to me, but I think we should wrap any exceptions thrown from `Activator.CreateInstance(nullableUnderlyingType)`.

```C#
            try
            {
                if (Nullable.GetUnderlyingType(type) is { } nullableUnderlyingType)
                {
                    // Activator.CreateInstance only returns null for Nullable<T> types.
                    return Activator.CreateInstance(nullableUnderlyingType)!;
                }

                return Activator.CreateInstance(type)!;
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException(SR.Format(SR.Error_FailedToActivate, type), ex);
            }
```

If we don't believe `Activator.CreateInstance only returns null for Nullable<T> types`, we should be checking the return values for both calls to `Activator.CreateInstance` to see if they're null.

Reply:
>  I think we should wrap any exceptions thrown from Activator.CreateInstance(nullableUnderlyingType).

Good call. I forgot that this can happen with default struct constructors.

> If we don't believe Activator.CreateInstance only returns null for Nullable<T> types, we should be checking the return values for both calls to Activator.CreateInstance to see if they're null.

@halter73 I'll defer to you on this. I vaguely recall reading a C# article years ago that said some COM objects could result in `null` returns from `Activator.CreateInstance`. No idea whether that's still true in .NET Core, whether we care if it is, or whether there are other edge cases like that to consider. Unfortunately, I can't find the original source. If we think explicitly handling `Nullable<T>` is sufficient to guarantee a non-null return, we can just use `!` as in your example.

Reply:
@madelson I tried to look if `Activator.CreateInstance` can return null for other cases but couldn't know if this is the case. Should we do the null check anyway to be in the safe side?

@halter73 what you think?

Reply:
> Should we do the null check anyway to be in the safe side?

I'm also unsure if `Activator.CreateInstance` can return null in other cases. It's the only scenario I see where [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.activator.createinstance?view=net-6.0#system-activator-createinstance(system-type)) say `CreateInstance` will return null, but I think null checking all created instances as the PR currently does (except for with well-known types like `List<T>`) makes sense.

Reply:
Thanks! Anything else for me to do here @tarekgh @halter73 ?